File: backend/src/main/java/com/programmers/pcquotation/domain/customer/entity/Customer.java
package com.programmers.pcquotation.domain.customer.entity;

import java.util.Collection;
import java.util.List;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;

import com.programmers.pcquotation.domain.comment.emtity.Comment;
import com.programmers.pcquotation.domain.member.entitiy.Member;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@Builder
@Entity
public class Customer implements Member {
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;
	@Column(unique = true)
	private String username;
	private String password;
	private String customerName;
	@Column(unique = true)
	private String email;
	private String verificationQuestion;
	private String verificationAnswer;
	@Column(unique = true)
	private String apiKey;

	@OneToMany(mappedBy = "customer", cascade = CascadeType.ALL, orphanRemoval = true)
	private List<Comment> comments;

	public Collection<? extends GrantedAuthority> getAuthorities() {
		return List.of("ROLE_CUSTOMER")
			.stream()
			.map(SimpleGrantedAuthority::new)
			.toList();
	}

	@Override
	public long getId() {
		return id;
	}

	@Override
	public String getUsername() {
		return username;
	}

	@Override
	public String getApiKey() {
		return apiKey;
	}

	@Override
	public String getPassword() {
		return password;
	}

}

File: backend/src/main/java/com/programmers/pcquotation/domain/customer/entity/Customer.kt
package com.programmers.pcquotation.domain.customer.entity

import com.programmers.pcquotation.domain.customer.dto.CustomerSignupRequest
import com.programmers.pcquotation.domain.member.entity.Member
import jakarta.persistence.*
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority

@Entity
class Customer : Member {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    override var id: Long? = null

    @Column(unique = true)
    override var username: String? = null
    override var password: String? = null
    var customerName: String = ""

    @Column(unique = true)
    var email: String? = null
    var verificationQuestion: String? = null
    var verificationAnswer: String? = null

    @Column(unique = true)
    override var apiKey: String? = null

    override var authorities: Collection<GrantedAuthority>? =
        listOf("ROLE_CUSTOMER").stream().map { role: String? -> SimpleGrantedAuthority(role) }.toList()

    constructor(
        username: String, password: String
    ) {
        this.username = username
        this.password = password
    }

    constructor(customerSignupRequest: CustomerSignupRequest, apiKey: String, encodingPassword: String) {
        this.username = customerSignupRequest.username
        this.password = encodingPassword
        this.customerName = customerSignupRequest.customerName
        this.email = customerSignupRequest.email
        this.verificationQuestion = customerSignupRequest.verificationQuestion
        this.verificationAnswer = customerSignupRequest.verificationAnswer
        this.apiKey = apiKey
    }

    constructor(username: String, password: String, customerName: String, email: String, apiKey: String) {
        this.username = username
        this.password = password
        this.customerName = customerName
        this.email = email
        this.apiKey = apiKey
    }

    constructor(
        username: String?,
        password: String?,
        customerName: String,
        email: String?,
        verificationQuestion: String?,
        verificationAnswer: String?
    ) {
        this.username = username
        this.password = password
        this.customerName = customerName
        this.email = email
        this.verificationQuestion = verificationQuestion
        this.verificationAnswer = verificationAnswer
    }

    constructor(
        id: Long?,
        username: String?,
        password: String?,
        customerName: String,
        email: String?,
        verificationQuestion: String?,
        verificationAnswer: String?,
        apiKey: String?
    ) {
        this.id = id
        this.username = username
        this.password = password
        this.customerName = customerName
        this.email = email
        this.verificationQuestion = verificationQuestion
        this.verificationAnswer = verificationAnswer
        this.apiKey = apiKey

    }

}